corehq/apps/reports/standard/users/reports.py

```diff
@@ -147,48 +150,65 @@ def rows(self):
+        except ObjectDoesNotExist:
+            return None
+        return location_object.display_name
+
+    def _user_history_row(self, record, domain, timezone):
+        return [
+            cached_user_id_to_username(record.user_id),
+            cached_user_id_to_username(record.changed_by),
+            _get_action_display(record.action),
+            record.details['changed_via'],
+            record.message,
+            self._user_history_details_cell(record.details['changes'], domain),
+            ServerTime(record.changed_at).user_time(timezone).ui_string(USER_DATETIME_FORMAT),
+        ]
 
+    def _user_history_details_cell(self, changes, domain):
+        def _html_list(changes, unstyled=True):
+            items = []
+            for key, value in changes.items():
+                if isinstance(value, dict):
+                    value = _html_list(value, unstyled=unstyled)
+                elif isinstance(value, list):
+                    value = format_html(", ".join(value))
+                else:
+                    value = format_html(str(value))
+                items.append("<li>{}: {}</li>".format(key, value))
+
+            class_attr = "class='list-unstyled'" if unstyled else ""
+            return mark_safe(f"<ul {class_attr}>{''.join(items)}</ul>")
+
+        properties = UserHistoryReport.get_primary_properties(domain)
+        properties.pop("user_data", None)
+        primary_changes = {}
+        all_changes = {}
 
-def _user_history_details_cell(changes, domain):
-    def _html_list(changes, unstyled=True):
-        items = []
         for key, value in changes.items():
-            if isinstance(value, dict):
-                value = _html_list(value, unstyled=unstyled)
-            elif isinstance(value, list):
-                value = format_html(", ".join(value))
-            else:
-                value = format_html(str(value))
-            items.append("<li>{}: {}</li>".format(key, value))
-
-        class_attr = "class='list-unstyled'" if unstyled else ""
-        return mark_safe(f"<ul {class_attr}>{''.join(items)}</ul>")
-
-    properties = UserHistoryReport.get_primary_properties(domain)
-    properties.pop("user_data", None)
-    primary_changes = {
-        properties.get(key, key): value for key, value in changes.items()
-        if key in properties
-    }
-    more_count = len(changes) - len(primary_changes)
-
-    return render_to_string("reports/standard/partials/user_history_changes.html", {
-        "primary_changes": _html_list(primary_changes) if primary_changes else None,
-        "all_changes": _html_list(changes, unstyled=False),
-        "more_count": more_count,
+            if key in properties:
+                if key in ('commcare_location_id', 'location_id'):
+                    value = self._get_location_name(value)
+                primary_changes[properties[key]] = value
+                all_changes[properties[key]] = value
+            if key == 'user_data':
```

Comment:
It looks like `all_changes` will end up both with a `user_data` key pointing to the full dict and also individual keys for each item in user data, is that intended?

Reply:
`user_data` is popped off of `properties` [here](https://github.com/dimagi/commcare-hq/pull/30330/files#:~:text=UserHistoryReport.get_primary_properties(domain)-,properties.pop(%22user_data%22%2C%20None),-primary_changes%20%3D%20%7B%7D) so I don't think we were getting a pointer to the full dict. I could have been missing something before, but `all_changes` should now just be getting the keys/values from `user_data`.

Reply:
I think the point here is also that a key from user_data will overrite a key already present in `all_changes`?

Reply:
would that look something like this:
`if key == 'user_data':
                for key, value in changes['user_data'].items():
                    all_changes[f"user data: {key}"] = value`

Reply:
(but with proper indentation)

Reply:
yeah, exactly